### PR TITLE
create log directory with control.xml and merged installation.xml

### PR DESCRIPTION
**Pull request by aschnell**
For https://trello.com/c/xKoN23o3/906-2-include-the-controlxml-to-be-collected-by-savey2logs.

### CODE REVIEW

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    # to the current one. Child classes can override the value. Max
+    # value is 99.
+    KEEP_OLD_LOG_DIRS = 3
+
+    # The name of the directory in /var/log/YaST2 to create, rotate
+    # and delete. Child classes should override the value.
+    NAME = "undefined".freeze
+
+    # Prepare (create, rotate and delete) the log directory.
+    def prepare
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: is separate call needed? Why not have it in initialize? then simple LogDirRotator.new will be enough instead of calling prepare explicitely.

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    end
+
+    # Write a file in the log directory.
+    #
+    # @param dest [String] basename of destination
+    # @param content [String] content to write
+    def write(dest, content)
+      return if Yast::Mode.test
+      begin
+        File.open(log_dir + "/" + dest, "w") { |file| file.write(content) }
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: It can be written as `File.write(log_dir + "/" + dest, content)` https://ruby-doc.org/core-2.5.0/IO.html#method-c-write

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    #   mv #{NAME}    #{NAME}-01
+    #
+    # This will NOT create any new log directory.
+    def rotate_log_dirs
+      return unless File.exist?(base_dir)
+      if installation?
+        kill_old_log_dirs([File.basename(log_dir)])
+      else
+        log_dirs = old_log_dirs.sort
+        keep_dirs = log_dirs.shift(self.class::KEEP_OLD_LOG_DIRS)
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: this won;t work when KEEP_OLD_LOG_DIRS will be bigger then 10 as it is string sorting. See this output of irb:
```
irb(main):004:0> a = []; 11.times { |i| a << "t#{i}" }; a.sort
=> ["t0", "t1", "t10", "t2", "t3", "t4", "t5", "t6", "t7", "t8", "t9"]
```

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    end
+
+    # Kill (recursively remove) old log directories.
+    #
+    # @param log_dirs [Array<String>] directory names (without path) to remove
+    def kill_old_log_dirs(log_dirs)
+      log_dirs.each do |dir|
+        next unless File.exist?(base_dir + "/" + dir)
+        log.info("Removing old log dir #{dir}")
+        FileUtils.remove_dir(base_dir + "/" + dir)
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: this will potentially raise exception.  I suggest to use `FileUtils.remove_dir(base_dir + "/" + dir, true)`. See https://ruby-doc.org/stdlib-2.4.1/libdoc/fileutils/rdoc/FileUtils.html#method-c-remove_dir

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    def write(dest, content)
+      return if Yast::Mode.test
+      begin
+        File.open(log_dir + "/" + dest, "w") { |file| file.write(content) }
+      rescue SystemCallError => e
+        log.error e.to_s
+      end
+    end
+
+    # Return a suitable name for the devicegraph log directory
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: here it mention devicegraph, which is probably not intended.

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    #
+    #   mv #{NAME}-02 #{NAME}-03
+    #   mv #{NAME}-01 #{NAME}-02
+    #   mv #{NAME}    #{NAME}-01
+    #
+    # @param old_name [String] old directory name (without path)
+    def rename_old_log_dir(old_name)
+      new_name =
+        if old_name =~ /[0-9]+$/
+          old_name.next
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: nice usage of next :+1:

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 1 above)
Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Yes, one might not need the prepare at other code locations. And the class is not a singleton.

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 3 above)
Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: The dirs are t-01, t-02, ... so it does work.

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 4 above)
Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Exceptions are logged as documented.

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    # Rename an old log directory according to this schema:
+    #
+    #   mv #{NAME}-02 #{NAME}-03
+    #   mv #{NAME}-01 #{NAME}-02
+    #   mv #{NAME}    #{NAME}-01
+    #
+    # @param old_name [String] old directory name (without path)
+    def rename_old_log_dir(old_name)
+      new_name =
+        if old_name =~ /[0-9]+$/
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: I suggest something like `/\d+\z/`. In ruby `$` in regexp means matching also endline. So potentianally if filepath contain newline[1] this won't work as intented. and `\d` is shorter then `[0-9]`. More about ruby regexp at https://ruby-doc.org/core-2.5.1/Regexp.html#class-Regexp-label-Regexp-23match+method

[1] https://superuser.com/questions/129519/which-file-systems-support-newlines-in-file-names

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+          old_name.next
+        else
+          old_name + "-01"
+        end
+      log.info("Rotating log dir #{old_name} to #{new_name}")
+      File.rename(base_dir + "/" + old_name, base_dir + "/" + new_name)
```
Comment (jreidinger) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: NP: this can potentionally can raise SystemCallError. But I do not know what is proper reaction in that case.

File: package/yast2.changes [hunk 9]
```diff
@@ -1,3 +1,9 @@
+-------------------------------------------------------------------
+Fri May 17 11:17:17 CEST 2019 - aschnell@suse.com
+
+- create log directory with control.xml and merged installation.xml
```
Comment (jreidinger) [inline package/yast2.changes]: I worry that it will need bsc, jira or fate number.

Comment (aschnell) [reply package/yast2.changes]: Not for factory (according to the openSUSE rules). For SP2 we can assume that it gets accepted together with other changes that do have a number.

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 8 above)
Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Exceptions are logged as documented.

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 7 above)
Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Why should the filename contain newline in the first place. This code is reused and has worked fine for some time.

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 3 above)
Comment (jreidinger) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: ah, right

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 7 above)
Comment (jreidinger) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Well, reason is security. It is not critical for me, but in general for security reasons it is recommented to use `\A\z` instead `^$` in ruby. The second one worked, but can be tricked to not do what expected.

File: library/general/src/lib/yast2/log_dir_rotator.rb [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    # value is 99.
+    KEEP_OLD_LOG_DIRS = 3
+
+    # The name of the directory in /var/log/YaST2 to create, rotate
+    # and delete. Child classes should override the value.
+    NAME = "undefined".freeze
+
+    # Prepare (create, rotate and delete) the log directory.
+    def prepare
+      return if Yast::Mode.test
```
Comment (imobachgs) [inline library/general/src/lib/yast2/log_dir_rotator.rb]: Why is this line needed? If you, for some reason, need to keep it, please document in the method's description.

Comment (jreidinger) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: my guess is that ProductControl is used in old testsuite and you want to skip this file modification in old testsuite.

Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Because otherwise the old testsuite fails. But I can also document that.

Comment (imobachgs) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: Thanks, that's what I wanted to know. Mentioning it would be fine so I do not ask the question again in the future :)

File: library/general/src/lib/yast2/log_dir_rotator.rb (hunk 7 above)
Comment (aschnell) [reply library/general/src/lib/yast2/log_dir_rotator.rb]: OK, then I suggest you create a bug report so that is also gets fixed in yast2-storage-ng.
